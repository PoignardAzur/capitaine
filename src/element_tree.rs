use crate::glue::GlobalEventCx;
use crate::widget_sequence::WidgetSequence;

pub trait ElementTree<ExplicitState> {
    type Event;
    type AggregateComponentState: Default;
    type BuildOutput: VirtualDom<
        ExplicitState,
        Event = Self::Event,
        AggregateComponentState = Self::AggregateComponentState,
    >;

    fn build(
        self,
        prev_state: Self::AggregateComponentState,
    ) -> (Self::BuildOutput, Self::AggregateComponentState);
}

// TODO - Include documentation about what a Virtual DOM is and where the name comes from.
pub trait VirtualDom<ParentComponentState> {
    type Event;
    type AggregateComponentState: Default;

    // TODO - Might be superfluous
    type DomState;

    type TargetWidgetSeq: WidgetSequence;

    // update_value is intended to enable memoize-style HOC
    // where instead of returning a vdom node, it returns
    // something along the lines of struct KeepEverythingAsItWas()
    // Ugh. I'm not explaining this well.
    fn update_value(&mut self, other: Self);

    fn init_tree(&self) -> (Self::TargetWidgetSeq, Self::DomState);

    fn apply_diff(
        &self,
        other: &Self,
        prev_state: Self::DomState,
        widget_seq: &mut Self::TargetWidgetSeq,
    ) -> Self::DomState;

    fn process_event(
        &self,
        explicit_state: &mut ParentComponentState,
        children_state: &mut Self::AggregateComponentState,
        dom_state: &mut Self::DomState,
        _cx: &mut GlobalEventCx,
    ) -> Option<Self::Event>;
}

/*
use crate::glue::{DruidAppData, GlobalEventCx, Id};

use crate::react_widgets::make_button;
use crate::react_widgets::{SingleWidget, WidgetList, WidgetSequence, WidgetTuple};

// TODO - refactor away WidgetPod
use druid::widget as druid_w;
use druid::WidgetPod;


use druid::kurbo::{Rect, Size};

use crate::flex2::FlexParams;
use crate::glue::DruidAppData;
use druid::widget::Button;
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, UpdateCtx,
    Widget, WidgetPod,
};
*/
