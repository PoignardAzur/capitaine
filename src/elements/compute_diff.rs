use std::fmt::Debug;
use tracing::{instrument, trace};

// TODO - Reduce allocations here
#[derive(Default, Clone, Debug, PartialOrd, PartialEq, Ord, Eq, Hash)]
pub struct ListMutationItem<Key> {
    pub index: usize,
    pub preserved_count_before: usize,
    pub removed_count: usize,
    pub inserted_keys: Vec<Key>,
}

#[derive(Default, Clone, Debug, PartialOrd, PartialEq, Ord, Eq, Hash)]
pub struct ListMutation<Key> {
    pub items: Vec<ListMutationItem<Key>>,
}

impl<Key> ListMutation<Key> {
    pub fn new(items: Vec<ListMutationItem<Key>>) -> Self {
        ListMutation { items }
    }
}

#[instrument(skip(old_list, new_list))]
fn find_next_common_item<Key: Eq + Debug, T1, T2>(
    old_list: &[(Key, T1)],
    new_list: &[(Key, T2)],
    cursor_old: usize,
    cursor_new: usize,
) -> Option<(usize, usize)> {
    for idx_old in cursor_old..old_list.len() {
        for idx_new in cursor_new..new_list.len() {
            if old_list[idx_old].0 == new_list[idx_new].0 {
                return Some((idx_old, idx_new));
            }
        }
    }
    None
}

#[instrument(skip(old_list, new_list))]
pub fn compute_diff<Key: Eq + Debug + Clone, T1, T2>(
    old_list: &[(Key, T1)],
    new_list: &[(Key, T2)],
) -> ListMutation<Key> {
    let mut list_mutation = ListMutation { items: vec![] };
    let mut cursor_old = 0usize;
    let mut cursor_new = 0usize;
    let mut preserved_count_before = 0;

    // TODO - trace keys
    trace!("Comparing arrays");

    while let Some((next_old, next_new)) =
        find_next_common_item(old_list, new_list, cursor_old, cursor_new)
    {
        // If we skipped over some items, then we add then to a new mutation
        if cursor_new != next_new || cursor_old != next_old {
            list_mutation.items.push(ListMutationItem {
                index: cursor_old,
                preserved_count_before,
                removed_count: next_old - cursor_old,
                inserted_keys: new_list[cursor_new..next_new]
                    .iter()
                    .map(|item| item.0.clone())
                    .collect(),
            });
            preserved_count_before = 1;
        } else {
            preserved_count_before += 1;
        }
        cursor_old = next_old + 1;
        cursor_new = next_new + 1;
    }

    list_mutation.items.push(ListMutationItem {
        index: cursor_old,
        preserved_count_before,
        removed_count: old_list.len() - cursor_old,
        inserted_keys: new_list[cursor_new..new_list.len()]
            .iter()
            .map(|item| item.0.clone())
            .collect(),
    });

    list_mutation
}

#[allow(dead_code)]
#[cfg(test)]
mod tests {
    use super::*;
    use test_env_log::test;

    fn make_list(keys: &[i32]) -> Vec<(i32, ())> {
        keys.iter().map(|key| (*key, ())).collect()
    }

    fn make_mut_tail(index: usize, preserved_count_before: usize) -> ListMutationItem<i32> {
        ListMutationItem {
            index,
            preserved_count_before,
            removed_count: 0,
            inserted_keys: vec![],
        }
    }

    // TODO - O(N * M) for N the list size and M the mutation count
    // Use better algo
    // TODO - Implement more generic version instead of copy-pasting everywhere
    fn apply_diff(old_list: &[(i32, ())], mutation: &ListMutation<i32>) -> Vec<(i32, ())> {
        let mut index_diff = 0_isize;
        let mut new_list: Vec<_> = old_list.into();

        // TODO - reserve

        for mutation_item in &mutation.items {
            let index = (mutation_item.index as isize + index_diff) as usize;
            let range = index..(index + mutation_item.removed_count);

            // Calling .last() runs the entire iterator
            let _ = new_list
                .splice(
                    range,
                    mutation_item
                        .inserted_keys
                        .iter()
                        .cloned()
                        .map(|key| (key, ())),
                )
                .last();

            index_diff += mutation_item.inserted_keys.len() as isize;
            index_diff -= mutation_item.removed_count as isize;
        }

        new_list
    }

    #[test]
    fn empty_lists() {
        let old_list = make_list(&[]);
        let new_list = make_list(&[]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![make_mut_tail(0, 0),]
            }
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn identical_lists() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 2, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![make_mut_tail(old_list.len(), 3),]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn inserted_one() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 2, 999, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 2,
                        preserved_count_before: 2,
                        removed_count: 0,
                        inserted_keys: vec![999],
                    },
                    make_mut_tail(old_list.len(), 1),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn inserted_one_beginning() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[999, 1, 2, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 0,
                        preserved_count_before: 0,
                        removed_count: 0,
                        inserted_keys: vec![999],
                    },
                    make_mut_tail(old_list.len(), 3),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn inserted_one_end() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 2, 3, 999]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![ListMutationItem {
                    index: 3,
                    preserved_count_before: 3,
                    removed_count: 0,
                    inserted_keys: vec![999],
                },]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn removed_one() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 1,
                        preserved_count_before: 1,
                        removed_count: 1,
                        inserted_keys: vec![],
                    },
                    make_mut_tail(old_list.len(), 1),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn removed_first() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[2, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 0,
                        preserved_count_before: 0,
                        removed_count: 1,
                        inserted_keys: vec![],
                    },
                    make_mut_tail(old_list.len(), 2),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn removed_last() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 2]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![ListMutationItem {
                    index: 2,
                    preserved_count_before: 2,
                    removed_count: 1,
                    inserted_keys: vec![],
                },]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn removed_all() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![ListMutationItem {
                    index: 0,
                    preserved_count_before: 0,
                    removed_count: 3,
                    inserted_keys: vec![],
                },]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn changed_one() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 999, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 1,
                        preserved_count_before: 1,
                        removed_count: 1,
                        inserted_keys: vec![999],
                    },
                    make_mut_tail(old_list.len(), 1),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn changed_first() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[999, 2, 3]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 0,
                        preserved_count_before: 0,
                        removed_count: 1,
                        inserted_keys: vec![999],
                    },
                    make_mut_tail(old_list.len(), 2),
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn changed_last() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[1, 2, 999]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![ListMutationItem {
                    index: 2,
                    preserved_count_before: 2,
                    removed_count: 1,
                    inserted_keys: vec![999],
                },]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn changed_all() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[101, 102, 103]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![ListMutationItem {
                    index: 0,
                    preserved_count_before: 0,
                    removed_count: 3,
                    inserted_keys: vec![101, 102, 103],
                },]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    #[test]
    pub fn swap_order() {
        let old_list = make_list(&[1, 2, 3]);
        let new_list = make_list(&[3, 2, 1]);

        let diff = compute_diff(&old_list, &new_list);
        assert_eq!(
            diff,
            ListMutation {
                items: vec![
                    ListMutationItem {
                        index: 0,
                        preserved_count_before: 0,
                        removed_count: 0,
                        inserted_keys: vec![3, 2],
                    },
                    ListMutationItem {
                        index: 1,
                        preserved_count_before: 1,
                        removed_count: 2,
                        inserted_keys: vec![],
                    },
                ]
            },
        );
        assert_eq!(apply_diff(&old_list, &diff), new_list);
    }

    // TODO
    // - Add more robust reconciliation tests with fuzzing
    // - Add a "multiple changes" test
}
