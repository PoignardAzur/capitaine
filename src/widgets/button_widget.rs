use crate::glue::Action;
use crate::glue::DruidAppData;
use crate::glue::Id;
use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;
use crate::widgets::flex::FlexParams;

use druid::kurbo::{Rect, Size};
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle,
    LifeCycleCtx, PaintCtx, UpdateCtx, WidgetPod,
};
use druid::widget::{Button, Click, ControllerHost};

pub struct ButtonWidget(
    pub WidgetPod<DruidAppData, ControllerHost<Button<DruidAppData>, Click<DruidAppData>>>,
    pub Id,
);

impl ButtonWidget {
    pub fn new(text: String, id: Id) -> Self {
        let button = Button::new(text)
            .on_click(move |_, data: &mut DruidAppData, _| data.queue_action(id, Action::Clicked));

        ButtonWidget(WidgetPod::new(button), id)
    }
}

impl FlexWidget for ButtonWidget {
    fn flex_params(&self) -> FlexParams {
        FlexParams {
            flex: 1.0,
            alignment: None,
        }
    }

    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut DruidAppData, env: &Env) {
        self.0.event(ctx, event, data, env);
    }

    fn lifecycle(
        &mut self,
        ctx: &mut LifeCycleCtx,
        event: &LifeCycle,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.lifecycle(ctx, event, data, env);
    }   fn update(
        &mut self,
        ctx: &mut UpdateCtx,
        _old_data: &DruidAppData,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.update(ctx, data, env);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &DruidAppData,
        env: &Env,
    ) -> Size {
        self.0.layout(ctx, bc, data, env)
    }

    fn paint_rect(&self) -> Rect {
        self.0.paint_rect()
    }

    fn set_layout_rect(&mut self, ctx: &mut LayoutCtx, data: &DruidAppData, env: &Env, rect: Rect) {
        self.0.set_layout_rect(ctx, data, env, rect)
    }

    fn layout_rect(&self) -> Rect {
        self.0.layout_rect()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, data: &DruidAppData, env: &Env) {
        self.0.paint(ctx, data, env);
    }
}

impl WidgetSequence for ButtonWidget {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        vec![self]
    }
}
