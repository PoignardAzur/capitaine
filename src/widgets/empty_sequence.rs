use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

pub struct EmptySequence;

impl WidgetSequence for EmptySequence {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        Vec::new()
    }
}
