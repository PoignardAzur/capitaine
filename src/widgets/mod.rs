pub mod empty_sequence;
pub mod flex_widget;
pub mod optional_widget;
pub mod single_widget;
pub mod widget_list;
pub mod widget_tuple;

pub use empty_sequence::*;
pub use flex_widget as flex;
pub use optional_widget::*;
pub use single_widget::*;
pub use widget_list::*;
pub use widget_tuple::*;
