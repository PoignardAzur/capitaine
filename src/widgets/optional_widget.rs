use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

use either::{Either, Left, Right};

// ---

impl<Child: WidgetSequence> WidgetSequence for Option<Child> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.iter_mut().flat_map(|child| child.widgets()).collect()
    }
}

impl<ChildLeft: WidgetSequence, ChildRight: WidgetSequence> WidgetSequence
    for Either<ChildLeft, ChildRight>
{
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        // TODO - use iterator chain instead?
        match self {
            Left(child_left) => child_left.widgets(),
            Right(child_right) => child_right.widgets(),
        }
    }
}
