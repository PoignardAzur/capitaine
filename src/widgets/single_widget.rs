use crate::glue::DruidAppData;
use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;
use crate::widgets::flex::FlexParams;
use druid::kurbo::{Rect, Size};
use druid::{
    BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, UpdateCtx,
    Widget, WidgetPod,
};

pub struct SingleWidget<W: Widget<DruidAppData>>(pub WidgetPod<DruidAppData, W>);

impl<W: Widget<DruidAppData>> SingleWidget<W> {
    pub fn new(widget: W) -> Self {
        SingleWidget(WidgetPod::new(widget))
    }
}

impl<W: Widget<DruidAppData>> FlexWidget for SingleWidget<W> {
    fn flex_params(&self) -> FlexParams {
        FlexParams {
            flex: 1.0,
            alignment: None,
        }
    }

    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut DruidAppData, env: &Env) {
        self.0.event(ctx, event, data, env);
    }

    fn lifecycle(
        &mut self,
        ctx: &mut LifeCycleCtx,
        event: &LifeCycle,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.lifecycle(ctx, event, data, env);
    }

    fn update(
        &mut self,
        ctx: &mut UpdateCtx,
        _old_data: &DruidAppData,
        data: &DruidAppData,
        env: &Env,
    ) {
        self.0.update(ctx, data, env);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &DruidAppData,
        env: &Env,
    ) -> Size {
        self.0.layout(ctx, bc, data, env)
    }

    fn paint_rect(&self) -> Rect {
        self.0.paint_rect()
    }

    fn set_layout_rect(&mut self, ctx: &mut LayoutCtx, data: &DruidAppData, env: &Env, rect: Rect) {
        self.0.set_layout_rect(ctx, data, env, rect)
    }

    fn layout_rect(&self) -> Rect {
        self.0.layout_rect()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, data: &DruidAppData, env: &Env) {
        self.0.paint(ctx, data, env);
    }
}

impl<W: Widget<DruidAppData>> WidgetSequence for SingleWidget<W> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        vec![self]
    }
}

use crate::glue::Action;
use crate::glue::Id;
use druid::widget::Button;
use druid::widget::Click;
use druid::widget::ControllerHost;
pub fn make_button(
    text: String,
    id: Id,
) -> SingleWidget<ControllerHost<Button<DruidAppData>, Click<DruidAppData>>> {
    let button = Button::new(text)
        .on_click(move |_, data: &mut DruidAppData, _| data.queue_action(id, Action::Clicked));

    SingleWidget(WidgetPod::new(button))
}
