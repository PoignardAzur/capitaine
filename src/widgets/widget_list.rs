use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

pub struct WidgetList<Child: WidgetSequence> {
    pub children: Vec<Child>,
}

impl<Child: WidgetSequence> WidgetSequence for WidgetList<Child> {
    fn widgets(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.children
            .iter_mut()
            .flat_map(|child| child.widgets())
            .collect()
    }
}
