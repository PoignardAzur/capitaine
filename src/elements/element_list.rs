use crate::element_tree::{Element, NoEvent, VirtualDom};
use crate::elements::compute_diff::compute_diff;
use crate::glue::GlobalEventCx;
use crate::widgets::WidgetList;

use crate::element_tree::ReconcileCtx;

use derivative::Derivative;
use either::{Left, Right};
use std::collections::VecDeque;
use tracing::{debug_span, info, instrument};

// TODO - Add arbitrary index types

#[derive(Derivative, PartialEq, Eq, Hash)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = "Child: Clone"))]
pub struct ElementList<Child: Element<CpState, CpEvent>, CpState = (), CpEvent = NoEvent> {
    pub children: Vec<(String, Child)>,
    pub _comp_state: std::marker::PhantomData<CpState>,
    pub _comp_event: std::marker::PhantomData<CpEvent>,
}

#[derive(Derivative, PartialEq, Eq, Hash)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = "Child: Clone"))]
pub struct ElementListData<Child: VirtualDom<CpState, CpEvent>, CpState = (), CpEvent = NoEvent> {
    pub children: Vec<(String, Child)>,
    pub _comp_state: std::marker::PhantomData<CpState>,
    pub _comp_event: std::marker::PhantomData<CpEvent>,
}

// ----

impl<CpState, CpEvent, Child: Element<CpState, CpEvent>> Element<CpState, CpEvent>
    for ElementList<Child, CpState, CpEvent>
{
    type Event = NoEvent;
    type AggregateChildrenState = Vec<(String, Child::AggregateChildrenState)>;
    type BuildOutput = ElementListData<Child::BuildOutput, CpState, CpEvent>;

    #[instrument(name = "List", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        // FIXE - Handle duplicate keys
        // TODO - Add special case when Child::AggregateChildrenState.sizeof() == 0

        let mutation = compute_diff(&prev_state, &self.children);
        let mut prev_state_or_default = prev_state;

        // TODO - reserve
        // TODO - O(N * M) for N the list size and M the mutation count
        // Use better algo
        let mut index_diff = 0_isize;
        for mutation_item in &mutation.items {
            let index = (mutation_item.index as isize + index_diff) as usize;
            let range = index..(index + mutation_item.removed_count);

            // Calling .last() runs the entire iterator, which performs the splice
            let _ = prev_state_or_default
                .splice(
                    range,
                    mutation_item
                        .inserted_keys
                        .iter()
                        .cloned()
                        .map(|key| (key, Default::default())),
                )
                .last();

            index_diff += mutation_item.inserted_keys.len() as isize;
            index_diff -= mutation_item.removed_count as isize;
        }

        let (children, new_state): (Vec<_>, Vec<_>) = self
            .children
            .into_iter()
            .zip(prev_state_or_default)
            .map(|((key, item), (_key, item_prev_state))| {
                let (new_item, new_state) = item.build(item_prev_state);
                ((key.clone(), new_item), (key, new_state))
            })
            .unzip();

        (
            ElementListData {
                children,
                ..Default::default()
            },
            new_state,
        )
    }
}

impl<CpState, CpEvent, Child: VirtualDom<CpState, CpEvent>> VirtualDom<CpState, CpEvent>
    for ElementListData<Child, CpState, CpEvent>
{
    type Event = NoEvent;
    type AggregateChildrenState = Vec<(String, Child::AggregateChildrenState)>;
    type TargetWidgetSeq = WidgetList<Child::TargetWidgetSeq>;

    #[instrument(name = "List", skip(self, other))]
    fn update_value(&mut self, other: Self) {
        *self = other;
    }

    #[instrument(name = "List", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        WidgetList {
            children: self
                .children
                .iter()
                .map(|(_key, elem)| elem.init_tree())
                .collect(),
        }
    }

    #[instrument(name = "List", skip(self, other, widget_seq, ctx))]
    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        let mutation = compute_diff(&other.children, &self.children);

        let mut prev_data: Vec<_> = other
            .children
            .iter()
            .zip(widget_seq.children.iter_mut())
            .map(Left)
            .collect();

        // TODO - reserve
        // TODO - O(N * M) for N the list size and M the mutation count
        // Use better algo
        // TODO - simplify
        let mut index_diff = 0_isize;
        for mutation_item in &mutation.items {
            let index = (mutation_item.index as isize + index_diff) as usize;
            let spliced_range = index..(index + mutation_item.removed_count);

            let new_range = index..(index + mutation_item.inserted_keys.len());

            // Calling .last() runs the entire iterator, which performs the splice
            let _ = prev_data
                .splice(spliced_range, self.children[new_range].iter().map(Right))
                .last();

            index_diff += mutation_item.inserted_keys.len() as isize;
            index_diff -= mutation_item.removed_count as isize;
        }

        let mut widgets_to_insert = VecDeque::new();
        for item in self.children.iter().zip(prev_data) {
            let (_key, child_data) = item.0;
            let child_prev_data = item.1;
            match child_prev_data {
                Left(prev_data) => {
                    let ((_key, child_prev_data), child_widget_seq) = prev_data;
                    child_data.reconcile(child_prev_data, child_widget_seq, ctx);
                }
                Right(new_data) => {
                    let (_key, child_data) = new_data;
                    let new_widget_seq = debug_span!("init_tree").in_scope(|| {
                        info!("creating child");
                        child_data.init_tree()
                    });
                    widgets_to_insert.push_back(new_widget_seq);
                }
            }
        }

        // TODO - reserve
        // TODO - O(N * M) for N the list size and M the mutation count
        // Use better algo
        // TODO - simplify
        let mut index_diff = 0_isize;
        for mutation_item in &mutation.items {
            let index = (mutation_item.index as isize + index_diff) as usize;
            let spliced_range = index..(index + mutation_item.removed_count);

            // Calling .last() runs the entire iterator, which performs the splice
            //.insert(index, new_widget_seq);
            let _ = widget_seq
                .children
                .splice(
                    spliced_range,
                    widgets_to_insert.drain(0..mutation_item.inserted_keys.len()),
                )
                .last();

            index_diff += mutation_item.inserted_keys.len() as isize;
            index_diff -= mutation_item.removed_count as isize;
        }
    }

    #[instrument(
        name = "List",
        skip(self, component_state, children_state, widget_seq, cx)
    )]
    fn process_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<CpEvent> {
        for child_data in self
            .children
            .iter()
            .zip(children_state)
            .zip(widget_seq.children.iter_mut())
        {
            let (_key, child) = child_data.0 .0;
            let child_comp_state = child_data.0 .1;
            let child_widget_seq = child_data.1;
            if let Some(event) = child.process_event(
                component_state,
                &mut child_comp_state.1,
                child_widget_seq,
                cx,
            ) {
                return Some(event);
            }
        }
        return None;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use crate::elements::label::{Label, LabelData};
    use crate::elements::{MockState, WithMockState};
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    fn new_label_list<State>(names: &[&str]) -> ElementList<Label<State>, State> {
        let children: Vec<_> = names
            .into_iter()
            .map(|name| (String::from(*name), Label::new(*name)))
            .collect();
        ElementList {
            children,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
        }
    }

    fn new_label_list_with_state<State>(
        names: &[&str],
    ) -> ElementList<WithMockState<Label<State>, State>, State> {
        let children: Vec<_> = names
            .iter()
            .map(|name| (String::from(*name), Label::new(*name).with_mock_state()))
            .collect();
        ElementList {
            children,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
        }
    }

    #[test]
    fn empty_list() {
        let list = new_label_list(&[]);
        let (list_data, _) = list.clone().build(Default::default());

        assert_debug_snapshot!(list);
        assert_debug_snapshot!(list_data);

        assign_empty_state_type(&list);
    }

    #[test]
    fn new_list() {
        let list = new_label_list(&["aaa", "bbb", "ccc"]);
        let (list_data, _) = list.clone().build(Default::default());

        assert_debug_snapshot!(list);
        assert_debug_snapshot!(list_data);

        assert_eq!(
            list_data,
            ElementListData {
                children: vec![
                    (String::from("aaa"), LabelData::new("aaa")),
                    (String::from("bbb"), LabelData::new("bbb")),
                    (String::from("ccc"), LabelData::new("ccc")),
                ],
                _comp_state: Default::default(),
                _comp_event: Default::default(),
            },
        );

        assign_empty_state_type(&list);
    }

    #[test]
    fn new_list_with_existing_state() {
        let list = new_label_list_with_state::<MockState>(&["aaa", "bbb", "ccc"]);
        let list_state = vec![
            (String::from("bbb"), (MockState::new("Foobar"), ())),
            (String::from("notfound"), (MockState::new("IAmError"), ())),
        ];
        let (_, new_list_state) = list.clone().build(list_state);

        assert_eq!(
            new_list_state,
            vec![
                (String::from("aaa"), (MockState::new("HelloWorld"), ())),
                (String::from("bbb"), (MockState::new("Foobar"), ())),
                (String::from("ccc"), (MockState::new("HelloWorld"), ())),
            ],
        );
    }

    // TODO
    // - Add constructor
    // - Widget test
    // - Event test
}
