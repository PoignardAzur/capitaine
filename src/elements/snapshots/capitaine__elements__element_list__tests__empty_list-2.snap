---
source: src/elements/element_list.rs
expression: list_data
---
ElementListData {
    children: [],
    _expl_state: PhantomData,
}
